src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_health_check_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsHealthCheckManager::GcsHealthCheckManager(
+    instrumented_io_context &io_service,
+    std::function<void(const NodeID &)> on_node_death_callback,
+    int64_t initial_delay_ms,
+    int64_t timeout_ms,
+    int64_t period_ms,
+    int64_t failure_threshold)
+    : io_service_(io_service),
+      on_node_death_callback_(on_node_death_callback),
+      initial_delay_ms_(initial_delay_ms),
+      timeout_ms_(timeout_ms),
+      period_ms_(period_ms),
+      failure_threshold_(failure_threshold) {}
+
+GcsHealthCheckManager::~GcsHealthCheckManager() {}
+
+void GcsHealthCheckManager::RemoveNode(const NodeID &node_id) {
+  io_service_.dispatch(
+      [this, node_id]() {
+        auto iter = health_check_contexts_.find(node_id);
+        if (iter == health_check_contexts_.end()) {
+          return;
+        }
+        health_check_contexts_.erase(iter);
+      },
+      "GcsHealthCheckManager::RemoveNode");
+}
+
+void GcsHealthCheckManager::FailNode(const NodeID &node_id) {
+  RAY_LOG(WARNING) << "Node " << node_id << " is dead because the health check failed.";
+  on_node_death_callback_(node_id);
+  health_check_contexts_.erase(node_id);
+}
+
+std::vector<NodeID> GcsHealthCheckManager::GetAllNodes() const {
+  std::vector<NodeID> nodes;
+  for (const auto &[node_id, _] : health_check_contexts_) {
+    nodes.emplace_back(node_id);
+  }
+  return nodes;
+}
+
+void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
+  using ::grpc::health::v1::HealthCheckResponse;
```

Comment:
IIUC, although you call timer_.cancel(), this method still can be executed if it was already queued. In this case, I think the stub_ can SIGSEGV because it is already GC'ed. I think we should always copy stopped to this block to avoid it being GC'ed and check it in the beginning. 

```
                timer_.async_wait([this, stopped](auto ec) {
                  if (ec != boost::asio::error::operation_aborted) {
                    StartHealthCheck(stopped);
                  }

void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck(stopped) {
    if (*stopped) {
        return; 
    }
```

Reply:
detail: https://github.com/ray-project/ray/pull/15070

Reply:
Ah, I think that's probably why node failure release test failed. Thanks for the catching!